# Data exploration and cleaning 

In [ ]:
import pandas as pd 


In [ ]:
data_categories=pd.read_csv("categories.csv",sep=';')

In [ ]:
data_categories.head()

category_id  ... parent_parent_parent_category
0          331  ...                           DPH
1          390  ...                      EPICERIE
2          837  ...                      LIQUIDES
3          419  ...                      EPICERIE
4          830  ...                      LIQUIDES

[5 rows x 8 columns]

In [ ]:
data_categories.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 335 entries, 0 to 334
Data columns (total 8 columns):
 #   Column                            Non-Null Count  Dtype 
---  ------                            --------------  ----- 
 0   category_id                       335 non-null    int64 
 1   category                          335 non-null    object
 2   parent_category_id                335 non-null    int64 
 3   parent_category                   335 non-null    object
 4   parent_parent_category_id         335 non-null    int64 
 5   parent_parent_category            335 non-null    object
 6   parent_parent_parent_category_id  335 non-null    int64 
 7   parent_parent_parent_category     335 non-null    object
dtypes: int64(4), object(4)
memory usage: 21.1+ KB


In [ ]:
data_product=pd.read_csv("product_carrefour.csv",sep=';')

In [ ]:
data_product.head()

product_id      brand  ... parent_parent_category  parent_parent_parent_category
0      242384  teisseire  ...           BRSA ET EAUX                       LIQUIDES
1      967177  livradois  ...               CREMERIE                            FLS
2      967177  livradois  ...               CREMERIE                            FLS
3      967269        NaN  ...  FRAIS NON LAITIERS LS                            FLS
4      968792    pamplie  ...               CREMERIE                            FLS

[5 rows x 8 columns]

In [ ]:
data_product.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 153343 entries, 0 to 153342
Data columns (total 8 columns):
 #   Column                         Non-Null Count   Dtype 
---  ------                         --------------   ----- 
 0   product_id                     153343 non-null  int64 
 1   brand                          152305 non-null  object
 2   name                           153343 non-null  object
 3   category_id                    153343 non-null  int64 
 4   category                       153343 non-null  object
 5   parent_category                153343 non-null  object
 6   parent_parent_category         153343 non-null  object
 7   parent_parent_parent_category  153343 non-null  object
dtypes: int64(2), object(6)
memory usage: 9.4+ MB


We'll only use data_product dataset later since it contains all the needed informations

In [ ]:
data_product["brand"].value_counts()

carrefour                         11811
refs.sans marque                   4468
crf cdm                            2389
nestle                             1397
carrefour bio                      1327
                                  ...  
labello classic                       1
a. legal                              1
brut de pomme                         1
les pâtisseries de miguelgorry        1
maison st geours                      1
Name: brand, Length: 5288, dtype: int64

In [ ]:
data_product["name"].value_counts()

Soda Coca-Cola                              83
Soda COCA-COLA                              47
Soda zero sucres Coca-Cola Zero             44
Lait demi-écrémé Candia                     41
Lait demi-écrémé CANDIA                     41
                                            ..
Chips piment doux et paprika Tyrrells        1
Sauce bourguignonne CARREFOUR                1
Confiture de figues ANDRE GUEPRATTE          1
Barres de céréales chocolat GRANY            1
Pain d'épices spécial foie gras Labeyrie     1
Name: name, Length: 98535, dtype: int64

In [ ]:
data_product[data_product["name"]=="Soda Coca-Cola"]

product_id  ... parent_parent_parent_category
3146       1277065  ...                      LIQUIDES
4119       1278569  ...                      LIQUIDES
4121       1278569  ...                      LIQUIDES
4205       1278659  ...                      LIQUIDES
5338       1280234  ...                      LIQUIDES
...            ...  ...                           ...
117064    15167861  ...                      LIQUIDES
118013    15445384  ...                      LIQUIDES
118015    15445384  ...                      LIQUIDES
120927    16437280  ...                      LIQUIDES
134011    20700189  ...                      LIQUIDES

[83 rows x 8 columns]

 we have some duplicated samples like in the exemple above of soda coca cola we ll remove all the duplicated product names from our dataset 

In [ ]:
def Lower(text):
   text = text.lower()
   return text

In [ ]:
data_product["name"]=data_product["name"].apply(Lower)
data_product["category"]=data_product["category"].apply(Lower)
data_product["parent_category"]=data_product["parent_category"].apply(Lower)
data_product["parent_parent_category"]=data_product["parent_parent_category"].apply(Lower)
data_product["parent_parent_parent_category"]=data_product["parent_parent_parent_category"].apply(Lower)

In [ ]:
data_product= data_product.drop_duplicates(subset = ["name"])

In [ ]:
data_product.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 70144 entries, 0 to 153341
Data columns (total 8 columns):
 #   Column                         Non-Null Count  Dtype 
---  ------                         --------------  ----- 
 0   product_id                     70144 non-null  int64 
 1   brand                          69249 non-null  object
 2   name                           70144 non-null  object
 3   category_id                    70144 non-null  int64 
 4   category                       70144 non-null  object
 5   parent_category                70144 non-null  object
 6   parent_parent_category         70144 non-null  object
 7   parent_parent_parent_category  70144 non-null  object
dtypes: int64(2), object(6)
memory usage: 4.8+ MB


In [ ]:
data_product.isnull().sum()

product_id                         0
brand                            895
name                               0
category_id                        0
category                           0
parent_category                    0
parent_parent_category             0
parent_parent_parent_category      0
dtype: int64

In [ ]:
data_brand_null=data_product[data_product["brand"].isnull()]

In [ ]:
data_brand_null.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 895 entries, 3 to 153253
Data columns (total 8 columns):
 #   Column                         Non-Null Count  Dtype 
---  ------                         --------------  ----- 
 0   product_id                     895 non-null    int64 
 1   brand                          0 non-null      object
 2   name                           895 non-null    object
 3   category_id                    895 non-null    int64 
 4   category                       895 non-null    object
 5   parent_category                895 non-null    object
 6   parent_parent_category         895 non-null    object
 7   parent_parent_parent_category  895 non-null    object
dtypes: int64(2), object(6)
memory usage: 62.9+ KB


As we can notice we have 895 null value in the brand column We will try to replace those null values

In [ ]:
data_brand_null["name"]

3                         quiche saumon courgettes grillées
5                                bonbons bergamote de nancy
7                                       essuie-tout compact
8                                       fromage bleu cendré
9                                       vin blanc au citron
                                ...                        
153196    ravioles fromages ail & fines herbes bio les a...
153208            rouille du pêcheur conserverie au bec fin
153212                        savon noir traditionnel zohra
153242          chocolat au lait le petit carre de chocolat
153253                 boisson melon nata de coco mogu mogu
Name: name, Length: 895, dtype: object

We cant find out the name of the brand from the product name that is why we ll proceed with another method

In [ ]:
data_brand_null["parent_parent_category"]=data_brand_null["parent_parent_category"].apply(Lower)

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


In [ ]:
data_brand_null.head()

product_id brand  ...    parent_parent_category  parent_parent_parent_category
3      967269   NaN  ...     frais non laitiers ls                            fls
5      968868   NaN  ...           epicerie sucree                       epicerie
7      970071   NaN  ...                   hygiene                            dph
8      970075   NaN  ...                  cremerie                            fls
9      970078   NaN  ...  spiritueux et champagnes                       liquides

[5 rows x 8 columns]

In [ ]:
data_brand_null= data_brand_null.drop_duplicates(subset = ["parent_parent_category"])

In [ ]:
data_brand_null.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 10 entries, 3 to 256
Data columns (total 8 columns):
 #   Column                         Non-Null Count  Dtype 
---  ------                         --------------  ----- 
 0   product_id                     10 non-null     int64 
 1   brand                          0 non-null      object
 2   name                           10 non-null     object
 3   category_id                    10 non-null     int64 
 4   category                       10 non-null     object
 5   parent_category                10 non-null     object
 6   parent_parent_category         10 non-null     object
 7   parent_parent_parent_category  10 non-null     object
dtypes: int64(2), object(6)
memory usage: 720.0+ bytes


In [ ]:
data_brand_null["parent_parent_category"].value_counts()

bieres et cidres            1
entretien                   1
epicerie salee              1
frais non laitiers ls       1
spiritueux et champagnes    1
cremerie                    1
brsa et eaux                1
hygiene                     1
surgeles glaces             1
epicerie sucree             1
Name: parent_parent_category, dtype: int64

In [ ]:
data_null=data_product[data_product["parent_parent_category"]=="surgeles glaces"]

In [ ]:
data_null["brand"].value_counts()

carrefour             769
findus                234
haagen-dazs           112
buitoni               108
marie                  96
                     ... 
of tov                  1
authentic asia          1
swedish glace           1
mexicano autentico      1
pilpa création          1
Name: brand, Length: 249, dtype: int64

since carrefour is the most featured brand in the parent_parent category surgeles glaces we will fill brand null columns with carrefour . We will use the same approach for the other null raws

In [ ]:
for i in data_product.index:
  if (data_product["parent_parent_category"][i]=="surgeles glaces" and pd.isnull(data_product["brand"][i])):
      data_product["brand"][i]="carrefour"

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  This is separate from the ipykernel package so we can avoid doing imports until


In [ ]:
data_null=data_product[data_product["parent_parent_category"]=="entretien"]
data_null["brand"].value_counts()

carrefour        526
air wick         158
glade            155
soupline         137
ariel            135
                ... 
compactor          1
mercurochrome      1
arelane            1
décolor stop       1
scotchgard         1
Name: brand, Length: 217, dtype: int64

In [ ]:
for i in data_product.index:
  if (data_product["parent_parent_category"][i]=="entretien" and pd.isnull(data_product["brand"][i])):
      data_product["brand"][i]="carrefour"

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  This is separate from the ipykernel package so we can avoid doing imports until


In [ ]:
data_null=data_product[data_product["parent_parent_category"]=="brsa et eaux"]
data_null["brand"].value_counts()

carrefour       255
tropicana        84
teisseire        84
pressade         75
joker            70
               ... 
oasis sirops      1
paris cola        1
sans              1
quezac            1
mont pele         1
Name: brand, Length: 284, dtype: int64

In [ ]:
for i in data_product.index:
  if (data_product["parent_parent_category"][i]=="brsa et eaux" and pd.isnull(data_product["brand"][i])):
      data_product["brand"][i]="carrefour"

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  This is separate from the ipykernel package so we can avoid doing imports until


In [ ]:
data_null=data_product[data_product["parent_parent_category"]=="frais non laitiers ls"]
data_null["brand"].value_counts()

carrefour               529
fleury michon           291
herta                   176
sodebo                  173
refs.sans marque        144
                       ... 
les tables normandes      1
choucrouterie wagner      1
salaisons dupoux          1
le malicieux              1
l'italie des pïf2tes      1
Name: brand, Length: 479, dtype: int64

In [ ]:
for i in data_product.index:
  if (data_product["parent_parent_category"][i]=="frais non laitiers ls" and pd.isnull(data_product["brand"][i])):
      data_product["brand"][i]="carrefour"

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  This is separate from the ipykernel package so we can avoid doing imports until


In [ ]:
data_null=data_product[data_product["parent_parent_category"]=="hygiene"]
data_null["brand"].value_counts()

carrefour               634
nivea                   476
l'oreal                 383
le petit marseillais    352
bourjois                327
                       ... 
gillette venus            1
saforelle                 1
venteo                    1
topicrem                  1
dermoexpertise            1
Name: brand, Length: 654, dtype: int64

In [ ]:
for i in data_product.index:
  if (data_product["parent_parent_category"][i]=="hygiene" and pd.isnull(data_product["brand"][i])):
      data_product["brand"][i]="carrefour"

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  This is separate from the ipykernel package so we can avoid doing imports until


In [ ]:
data_null=data_product[data_product["parent_parent_category"]=="cremerie"]
data_null["brand"].value_counts()

carrefour               683
refs.sans marque        241
yoplait                 218
andros                  214
la laitiere             185
                       ... 
climont                   1
cabreignac                1
crest                     1
l'oeufs de nos villa      1
et aldia                  1
Name: brand, Length: 680, dtype: int64

In [ ]:
for i in data_product.index:
  if (data_product["parent_parent_category"][i]=="cremerie" and pd.isnull(data_product["brand"][i])):
      data_product["brand"][i]="carrefour"

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  This is separate from the ipykernel package so we can avoid doing imports until


In [ ]:
data_null=data_product[data_product["parent_parent_category"]=="epicerie salee"]
data_null["brand"].value_counts()


carrefour              1376
bledina                 910
knorr                   369
refs.sans marque        354
nestle                  346
                       ... 
bob martin                1
moulin saint-michel       1
baff                      1
twisty                    1
esmeralda                 1
Name: brand, Length: 763, dtype: int64

In [ ]:
for i in data_product.index:
  if (data_product["parent_parent_category"][i]=="epicerie salee" and pd.isnull(data_product["brand"][i])):
      data_product["brand"][i]="carrefour"

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  This is separate from the ipykernel package so we can avoid doing imports until


In [ ]:
data_null=data_product[data_product["parent_parent_category"]=="epicerie sucree"]
data_null["brand"].value_counts()


carrefour                 1291
vahine                     301
lindt                      288
nestle                     237
cote d'or                  190
                          ... 
camel                        1
terres et céréales bio       1
café poitau                  1
biscuits arena               1
novi                         1
Name: brand, Length: 860, dtype: int64

In [ ]:
for i in data_product.index:
  if (data_product["parent_parent_category"][i]=="epicerie sucree" and pd.isnull(data_product["brand"][i])):
      data_product["brand"][i]="carrefour"

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  This is separate from the ipykernel package so we can avoid doing imports until


In [ ]:
data_null=data_product[data_product["parent_parent_category"]=="bieres et cidres"]
data_null["brand"].value_counts()

refs.sans marque    96
leffe               44
desperados          32
1664                31
grimbergen          27
                    ..
singha               1
rochefort            1
eku 28               1
estribos             1
g de goudale         1
Name: brand, Length: 190, dtype: int64

In [ ]:
for i in data_product.index:
  if (data_product["parent_parent_category"][i]=="bieres et cidres" and pd.isnull(data_product["brand"][i])):
      data_product["brand"][i]="refs.sans marque"

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  This is separate from the ipykernel package so we can avoid doing imports until


In [ ]:
data_null=data_product[data_product["parent_parent_category"]=="spiritueux et champagnes"]
data_null["brand"].value_counts()

refs.sans marque    148
carrefour           121
jack daniels         24
freixenet            24
aberlour             23
                   ... 
malesan               1
talisker skye         1
rivarose              1
the balvenie          1
prince royer          1
Name: brand, Length: 646, dtype: int64

In [ ]:
for i in data_product.index:
  if (data_product["parent_parent_category"][i]=="spiritueux et champagnes" and pd.isnull(data_product["brand"][i])):
      data_product["brand"][i]="refs.sans marque"

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  This is separate from the ipykernel package so we can avoid doing imports until


In [ ]:
data_product.isnull().sum()

product_id                       0
brand                            0
name                             0
category_id                      0
category                         0
parent_category                  0
parent_parent_category           0
parent_parent_parent_category    0
dtype: int64

we finally replaced all the null raws with the respective values

In [ ]:
data_product["brand"]=data_product["brand"].apply(Lower)


In [ ]:
data_product.head()

product_id      brand  ... parent_parent_category  parent_parent_parent_category
0      242384  teisseire  ...           brsa et eaux                       liquides
1      967177  livradois  ...               cremerie                            fls
2      967177  livradois  ...               cremerie                            fls
3      967269  carrefour  ...  frais non laitiers ls                            fls
4      968792    pamplie  ...               cremerie                            fls

[5 rows x 8 columns]

In [ ]:
data_product=data_product.drop(["product_id","category_id"],axis=1)

In [ ]:
data_product["description"]=data_product["brand"]+" "+data_product["category"]+" "+data_product["parent_category"]+" "+data_product["parent_parent_category"]+" "+data_product["parent_parent_parent_category"]

In [ ]:
data_product=data_product.drop(["brand","category","parent_category","parent_parent_category","parent_parent_parent_category"],axis=1)

In [ ]:
data_product.head()

name                                        description
0  boisson aux fruits multivitaminé teisseire  teisseire boissons aux fruits plates brsa non ...
1                                     chabrol  livradois fromages de specialite fromages ls c...
2                                 chabrol igp  livradois fromages de specialite fromages ls c...
3           quiche saumon courgettes grillées  carrefour quiche/tarte salee/tourte rf traiteu...
4                       beurre extra-fin doux     pamplie beurres beurre oeufs lait cremerie fls

In [ ]:
data_product.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 70144 entries, 0 to 153341
Data columns (total 2 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   name         70144 non-null  object
 1   description  70144 non-null  object
dtypes: object(2)
memory usage: 4.1+ MB


# Data Preprocessing

In [ ]:
data_product["description"][3]

'carrefour quiche/tarte salee/tourte rf traiteur ls frais non laitiers ls fls'

We will preprocess our description text: we will remove the ponctuation,stopwords ....

In [ ]:
from nltk.corpus import stopwords
from nltk import word_tokenize
import nltk
nltk.download('stopwords')
STOPWORDS = set(stopwords.words('french'))
import re
import spacy
import string 

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.


In [ ]:
def clean_text(text):
    
    text = re.sub('\[.*?\]', ' ', text)
    text = re.sub('[%s]' % re.escape(string.punctuation), ' ', text)
    text = re.sub('\w*\d\w*', ' ', text)
    text = re.sub('\n', ' ', text)
    text = ' '.join(word for word in text.split() if word not in STOPWORDS)
    return text

In [ ]:
data_product["description"]=data_product["description"].apply(clean_text)

In [ ]:
data_product["description"][3]

'carrefour quiche tarte salee tourte rf traiteur ls frais non laitiers ls fls'

In [ ]:
data_product.to_csv("data_prepocessed.csv")

Our data is finally preprocessed we move next to the first approach 

# First Approach using language model

---



In the first approach we use the pretrained model bert to compute the word vectors of each product description then we define cosine similarity as a similarity metrics as it is the most suitable and accurate one for our use case.
Finally we define a function that takes 3 products names, calculates the similarity between the products and returns the most similar ones 

In [ ]:
import pandas as pd 
data_product=pd.read_csv("data_prepocessed.csv")

**Text embeddings**





In [ ]:
!pip install sentence-transformers

     |████████████████████████████████| 85 kB 2.8 MB/s 
     |████████████████████████████████| 2.8 MB 11.7 MB/s 
     |████████████████████████████████| 1.2 MB 46.7 MB/s 
     |████████████████████████████████| 52 kB 1.5 MB/s 
     |████████████████████████████████| 895 kB 57.7 MB/s 
     |████████████████████████████████| 636 kB 67.3 MB/s 
     |████████████████████████████████| 3.3 MB 42.4 MB/s 
  Created wheel for sentence-transformers: filename=sentence_transformers-2.0.0-py3-none-any.whl size=126710 sha256=3744e1824bc99ff7d4b56607eb02aff201d444ecf32b532f5fafd08ead9271c9
  Stored in directory: /root/.cache/pip/wheels/d1/c1/0f/faafd427f705c4b012274ba60d9a91d75830306811e1355293
Successfully built sentence-transformers
  Attempting uninstall: pyyaml
    Found existing installation: PyYAML 3.13
    Uninstalling PyYAML-3.13:
      Successfully uninstalled PyYAML-3.13


In [ ]:
from sentence_transformers import SentenceTransformer

model = SentenceTransformer('bert-base-nli-mean-tokens')

Downloading:   0%|          | 0.00/391 [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/3.95k [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/2.00 [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/625 [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/122 [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/229 [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/438M [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/53.0 [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/112 [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/466k [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/399 [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/232k [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/190 [00:00<?, ?B/s]

In [ ]:
description=list(data_product["description"])

In [ ]:
description_embeddings = model.encode(description)


In [ ]:
description_embeddings

In [ ]:
description_embeddings.shape

since it takes a long time to train the model we save description_embeddings

In [ ]:
import pickle 
with open("embeddings.pickle","wb") as f :
  pickle.dump(description_embeddings,f)

In [ ]:
import pickle
pickle_in=open("embeddings.pickle","rb")
description_embeddings=pickle.load(pickle_in)

In [ ]:
description_embeddings.shape

we are going to define a function that takes in  3 product names as an input and outputs the most two similar ones  . Firstly, for this, we need a reverse mapping of product names and DataFrame indices. In other words, we need a mechanism to identify the index of a product name in our DataFrame, given its name

In [ ]:
ls=[]
for i in range (70144):
  ls.append(i) 

In [ ]:
import numpy as np
tab=np.array(ls)

In [ ]:
indices = pd.Series(tab, index=data_product['name']).drop_duplicates()


In [ ]:
indices

name
boisson aux fruits multivitaminé teisseire                                 0
chabrol                                                                    1
chabrol igp                                                                2
quiche saumon courgettes grillées                                          3
beurre extra-fin doux                                                      4
                                                                       ...  
piment d'espelette aop albert menes                                    70139
noisettes décortiquées bio les fadas                                   70140
bière blondes daytime lagunitas                                        70141
complément alimentaire vitamine c 500 laboratoires juvamine            70142
complément alimentaire vitamine c & magnésium laboratoires juvamine    70143
Length: 70144, dtype: int64

In [ ]:
import pickle 
with open("indice.pickle","wb") as f :
  pickle.dump(indices,f)

**Similarity Metric**

In [ ]:
from sklearn.metrics.pairwise import cosine_similarity


In [ ]:
def Distance (product_a,product_b,product_c):
  a=indices[product_a]
  b=indices[product_b]
  c=indices[product_c]
  d1=cosine_similarity(
    [description_embeddings[a]],
    [description_embeddings[b]])
  d2=cosine_similarity(
    [description_embeddings[a]],
    [description_embeddings[c]])
  if (d1>d2):
    print(product_a," est sémantiquement plus proche de ",product_b,"que ",product_c)
    print("D1:{}".format(d1))
    print("D2:{}".format(d2))

  elif (d1<d2):
    print(product_a," est sémantiquement plus proche de ",product_c,"que ",product_b)
    print("D1:{}".format(d1))
    print("D2:{}".format(d2))
  else:
    print(product_a," est sémantiquement égal à ",product_c,"et ",product_b)
    print("D1:{}".format(d1))
    print("D2:{}".format(d2))

 
  

Next we try an exemple . we will find which one between "bière blondes daytime lagunitas" and "beurre extra-fin doux" is the most similar to "boisson aux fruits multivitaminé teisseire"

In [ ]:
Distance("boisson aux fruits multivitaminé teisseire","bière blondes daytime lagunitas","beurre extra-fin doux")

# Second Appoach using Doc2vec

In the second Approach we will try using  doc2vec which is  is an extension of word2vec to the phrase-, sentence-, and document-level.We will consider each description of a given product name as document and we will try to find similarity between the products. First we will build our vocabulary of our dataset in order  fine tune the model.
The model can then return the similarity between two documents(product's description) 

In [ ]:
!pip install gensim


In [ ]:
import nltk
nltk.download('punkt')

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.


True

In [ ]:
import pandas as pd 
import pickle
data_product=pd.read_csv("data_prepocessed.csv")
pickle_in=open("indice.pickle","rb")
indices=pickle.load(pickle_in)

In [ ]:
data_product.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 70144 entries, 0 to 70143
Data columns (total 3 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   Unnamed: 0   70144 non-null  int64 
 1   name         70144 non-null  object
 2   description  70144 non-null  object
dtypes: int64(1), object(2)
memory usage: 1.6+ MB


In [ ]:
data=data_product.description

In [ ]:
import gensim
from gensim.models.doc2vec import Doc2Vec, TaggedDocument
from nltk.tokenize import word_tokenize
from gensim.models.doc2vec import Doc2Vec

In [ ]:
tagged_data = [TaggedDocument(words=word_tokenize(_d.lower()), tags=[str(i)]) for i, _d in enumerate(data)]

In [ ]:
tagged_data[0]

TaggedDocument(words=['teisseire', 'boissons', 'fruits', 'plates', 'brsa', 'non', 'gazeuses', 'brsa', 'eaux', 'liquides'], tags=['0'])

In [ ]:
tagged_data[1]

TaggedDocument(words=['livradois', 'fromages', 'specialite', 'fromages', 'ls', 'cremerie', 'fls'], tags=['1'])

We can access to each document through its tag exp (tags=['1'])





In [ ]:
print(tagged_data)


IOPub data rate exceeded.
The notebook server will temporarily stop sending output
to the client in order to avoid crashing it.
To change this limit, set the config variable
`--NotebookApp.iopub_data_rate_limit`.

Current values:
NotebookApp.iopub_data_rate_limit=1000000.0 (bytes/sec)
NotebookApp.rate_limit_window=3.0 (secs)



We load the model

In [ ]:
model = gensim.models.doc2vec.Doc2Vec(vector_size=30, min_count=2, epochs=20)

We build our dataset vocab

In [ ]:
model.build_vocab(tagged_data)


then we train the model

In [ ]:
model.train(tagged_data, total_examples=model.corpus_count, epochs=20)

In [ ]:
model.save("product.model")


In [ ]:
model = Doc2Vec.load("product.model")


We give an exemple of how the model can calculate the similarity of two products descriptions given the product tag

In [ ]:
model.docvecs.similarity('0','1')

-0.02351066

next we  define a function that takes in  3 product names as an input and outputs the most two similar ones  . 

In [ ]:
def Distance (product_a,product_b,product_c):
  a=indices[product_a]
  b=indices[product_b]
  c=indices[product_c]
  a=str(a)
  b=str(b)
  c=str(c)
  d1=model.docvecs.similarity(a,b)
  d2=model.docvecs.similarity(a,c)
  if (d1>d2):
    print(product_a," est sémantiquement plus proche de ",product_b,"que ",product_c)
    print("D1:{}".format(d1))
    print("D2:{}".format(d2))

  elif (d1<d2):
    print(product_a," est sémantiquement plus proche de ",product_c,"que ",product_b)
    print("D1:{}".format(d1))
    print("D2:{}".format(d2))
  else:
    print(product_a," est sémantiquement égal à ",product_c,"et ",product_b)
    print("D1:{}".format(d1))
    print("D2:{}".format(d2))

In [ ]:
Distance("boisson aux fruits multivitaminé teisseire","bière blondes daytime lagunitas","beurre extra-fin doux")

boisson aux fruits multivitaminé teisseire  est sémantiquement plus proche de  bière blondes daytime lagunitas que  beurre extra-fin doux
D1:0.6232568621635437
D2:0.5121594667434692


# Third Appoach using google universal encoder

The third approach consists of fine tuning google universal encoder  with our dataset. then define a cosine similarity layer after sentence encodings.I didnt finish this approach because i dont know how to train the neural architecture and there isnt much ressource about it  (this approach is more about buiding the architecture from scratch)

In [ ]:

#we load universal encoder
import tensorflow as tf
import tensorflow_hub as hub
module_url = "https://tfhub.dev/google/universal-sentence-encoder-large/3"
# Import the Universal Sentence Encoder's TF Hub module
embed = hub.load(module_url)

In [ ]:
def UniversalEmbedding(x):
    return embed(tf.squeeze(tf.cast(x, tf.string)), 
    	signature="default", as_dict=True)["default"]

In [ ]:
input_text = layers.Input(shape=(1,), dtype=tf.string)
embedding = layers.Lambda(UniversalEmbedding,
	output_shape=(embed_size,))(input_text)
dense = keras.layers.Dot(axes, normalize=True)(embedding) #cosine similarity layer
pred = layers.Dense(1, activation='sigmoid')(dense)
model = Model(inputs=[input_text], outputs=pred)
model.compile(loss=  # don t know the loss function nor the mertics nor the optimizer